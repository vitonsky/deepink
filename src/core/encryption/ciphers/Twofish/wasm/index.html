<script src="./out/twofish.js"></script>
<script>
	console.log("Init...");
	TwofishModule().then((mod) => {
		console.log("Module is loaded", mod)
		globalThis.Module = mod;
	})
</script>