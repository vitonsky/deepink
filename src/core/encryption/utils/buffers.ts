/**
 * Join buffers into a new buffer
 */
export const joinBuffers = (buffers: (ArrayBufferLike | Buffer | TypedArray)[]) => {
	const bufferLen = buffers.reduce((len, buffer) => len + buffer.byteLength, 0);
	const resultBuffer = new Uint8Array(bufferLen);

	let offset = 0;
	for (const buffer of buffers) {
		// We may receive a TypedArray instead of ArrayBuffer, so we have to get a buffer
		const rawBuffer = ArrayBuffer.isView(buffer) ? buffer.buffer : buffer;
		resultBuffer.set(new Uint8Array(rawBuffer), offset);
		offset += rawBuffer.byteLength;
	}

	return resultBuffer.buffer;
};

/**
 * Fill buffer with paddings to ensure buffer size multiple 16
 * Before use original data you have to remove padding
 */
export function fillBuffer(buffer: Uint8Array, blockSize = 16): [Uint8Array, number] {
	const padding = Math.ceil(buffer.length / blockSize) * blockSize - buffer.length;
	if (padding === 0) return [buffer, 0];

	// Create new buffer with padding
	const out = new Uint8Array(buffer.length + padding);
	out.set(buffer);

	return [out, padding];
}

/**
 * Buffer view to manipulate buffer slices
 */
export class BufferView {
	private readonly buffer;
	constructor(buffer: ArrayBuffer) {
		this.buffer = buffer;
	}

	public getBytes(offset: number = 0, end?: number) {
		return this.buffer.slice(offset, end);
	}

	public setBytes(buffer: ArrayBuffer, offset: number = 0) {
		if (offset > this.buffer.byteLength)
			throw new RangeError('Offset out of buffer size');
		if (offset + buffer.byteLength > this.buffer.byteLength)
			throw new RangeError(
				'Buffer size with current offset will out of buffer size',
			);

		const srcBufferView = new Uint8Array(buffer);
		const targetBufferView = new Uint8Array(this.buffer);
		for (let srcOffset = 0; srcOffset < buffer.byteLength; srcOffset++) {
			targetBufferView[offset + srcOffset] = srcBufferView[srcOffset];
		}
	}
}

type TypedArray =
	| Int8Array
	| Uint8Array
	| Uint8ClampedArray
	| Int16Array
	| Uint16Array
	| Int32Array
	| Uint32Array
	| Float32Array
	| Float64Array
	| BigInt64Array
	| BigUint64Array;

/**
 * Receive any buffer and returns a converted transferable buffer and codec for revert converting
 *
 * Useful to convert buffer for transfer and then convert received buffer back.
 *
 * @param buffer
 * @returns [buffer, convertor]
 */
export const convertBufferToTransferable = <
	T extends ArrayBufferLike | Buffer | TypedArray,
>(
	buffer: T,
): [ArrayBuffer, (buffer: ArrayBuffer) => T] => {
	if (globalThis.Buffer !== undefined && buffer instanceof Buffer) {
		return [
			buffer.buffer as ArrayBuffer,
			(buffer: ArrayBuffer) => {
				return Buffer.from(buffer) as T;
			},
		];
	}

	// Typed arrays
	const typedArray = [
		Int8Array,
		Uint8Array,
		Uint8ClampedArray,
		Int16Array,
		Uint16Array,
		Int32Array,
		Uint32Array,
		Float32Array,
		Float64Array,
		BigInt64Array,
		BigUint64Array,
	].find((proto) => buffer instanceof proto);
	if (typedArray && buffer instanceof typedArray) {
		return [
			(buffer as Uint8Array).buffer as ArrayBuffer,
			(buffer: ArrayBuffer) => {
				return new typedArray(buffer) as T;
			},
		];
	}

	return [
		buffer as ArrayBuffer,
		(buffer: ArrayBuffer) => {
			return buffer as T;
		},
	];
};
