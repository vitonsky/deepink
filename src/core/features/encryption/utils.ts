import { ENCRYPTION_ALGORITHM, ENCRYPTION_ALGORITHM_LIST } from './algorithms';

export const joinAlgorithms = (algorithms: ENCRYPTION_ALGORITHM[]) => {
	return algorithms.join('-');
};

export function isValidAlgorithm(algorithm: string): algorithm is ENCRYPTION_ALGORITHM {
	return Object.values(ENCRYPTION_ALGORITHM).includes(
		algorithm as ENCRYPTION_ALGORITHM,
	);
}

export function parseAlgorithms(algorithms: string) {
	return algorithms.split('-').map((name) => {
		if (!isValidAlgorithm(name)) {
			throw new Error(
				`Unsupported encryption algorithm: ${algorithms}. Supported: ${ENCRYPTION_ALGORITHM_LIST.join(
					', ',
				)}`,
			);
		}
		return name;
	});
}
