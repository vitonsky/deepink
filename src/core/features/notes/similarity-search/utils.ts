export function cosineSimilarity(a: number[], b: number[]) {
	if (a.length !== b.length) throw new Error('Vectors must be same length');
	let dot = 0;
	let normA = 0;
	let normB = 0;
	for (let i = 0; i < a.length; i++) {
		dot += a[i] * b[i];
		normA += a[i] * a[i];
		normB += b[i] * b[i];
	}
	return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}
