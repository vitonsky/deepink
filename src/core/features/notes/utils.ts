import { INoteContent } from '@core/features/notes/index';

export const getNoteTitle = (note: INoteContent, limit?: number) => {
	let title = (note.title || note.text).trim();
	if (limit) {
		title = title.slice(0, limit).trim();
	}

	return title || 'Empty note';
};
