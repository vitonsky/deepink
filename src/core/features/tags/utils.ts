import { IResolvedTag } from '.';

// TODO: remove unused code
export const isTagsArray = (data: unknown): data is string[] =>
	Array.isArray(data) && data.every((item) => typeof item === 'string');

export const findParentTag = (resolvedTagName: string, tags: IResolvedTag[]) => {
	let parent: IResolvedTag | null = null;
	for (const tag of tags) {
		if (!resolvedTagName.startsWith(tag.resolvedName)) continue;

		// Remember only those parent who have more segments
		if (parent === null || parent.resolvedName.length < tag.resolvedName.length) {
			parent = tag;
		}
	}

	return parent;
};
