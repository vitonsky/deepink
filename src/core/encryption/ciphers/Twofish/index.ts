import { encrypt, makeSession } from 'twofish-ts';

import { CTRCipherMode } from '../../cipherModes/CTRCipherMode';
import { BufferView, fillBuffer, joinBuffers } from '../../utils/buffers';

import { HeaderView, IEncryptionProcessor, RandomBytesGenerator } from '../..';

const blockSize = 16;

export type TwofishBufferHeaderStruct = {
	padding: number;
	iv: ArrayBuffer;
};

export class TwofishBufferHeader implements HeaderView<TwofishBufferHeaderStruct> {
	public readonly bufferSize = 128;
	private readonly offsets = {
		padding: 0,
		iv: 1,
	};

	private readonly ivSize;
	constructor(ivSize: number) {
		this.ivSize = ivSize;
	}

	public createBuffer(data: TwofishBufferHeaderStruct): ArrayBuffer {
		const buffer = new ArrayBuffer(this.bufferSize);
		const view = new DataView(buffer, 0);
		const rawView = new BufferView(buffer);

		view.setUint8(this.offsets.padding, data.padding);
		rawView.setBytes(data.iv, this.offsets.iv);

		return buffer;
	}

	public readBuffer(buffer: ArrayBuffer): TwofishBufferHeaderStruct {
		if (buffer.byteLength < this.bufferSize)
			throw new TypeError('Header buffer have too small size');

		const view = new DataView(buffer, 0, this.bufferSize);
		const rawView = new BufferView(buffer);

		return {
			padding: view.getUint8(this.offsets.padding),
			iv: rawView.getBytes(this.offsets.iv, this.offsets.iv + this.ivSize),
		};
	}
}

/**
 * Inner util to transform block with cipher
 * Util creates a new buffer instead of mutate original buffer.
 */
function transformBuffer(
	buffer: Uint8Array,
	transformBlock: (offset: number, input: Uint8Array, output: Uint8Array) => void,
) {
	const out = new Uint8Array(buffer.length);
	for (let offset = 0; offset < buffer.length; offset += blockSize) {
		transformBlock(offset, buffer, out);
	}

	return out;
}

// TODO: implement GCM mode https://en.wikipedia.org/wiki/Galois/Counter_Mode
/**
 * Twofish cipher implementation
 */
export class TwofishCTRCipher implements IEncryptionProcessor {
	private readonly ivSize = 96;

	private readonly key;
	private readonly randomBytesGenerator: RandomBytesGenerator;
	private readonly header;
	private readonly ctrCipher;
	constructor(cipher: Uint8Array, randomBytesGenerator: RandomBytesGenerator) {
		this.randomBytesGenerator = randomBytesGenerator;

		this.key = makeSession(cipher);
		this.header = new TwofishBufferHeader(this.ivSize);
		this.ctrCipher = new CTRCipherMode(this.encryptBuffer);
	}

	private encryptBuffer = async (buffer: ArrayBuffer) => {
		return transformBuffer(new Uint8Array(buffer), (offset, input, output) => {
			encrypt(input, offset, output, offset, this.key);
		}).buffer;
	};

	public async encrypt(buffer: ArrayBuffer) {
		const [bufferView, padding] = fillBuffer(new Uint8Array(buffer));

		const iv = this.randomBytesGenerator(this.ivSize);
		const encryptedBuffer = await this.ctrCipher.encrypt(
			bufferView.buffer as ArrayBuffer,
			iv,
		);

		const header = this.header.createBuffer({ padding, iv });
		return joinBuffers([header, encryptedBuffer]);
	}

	public async decrypt(buffer: ArrayBuffer) {
		const { padding, iv } = this.header.readBuffer(buffer);

		const encryptedBuffer = new Uint8Array(buffer.slice(this.header.bufferSize));
		const decryptedBufferWithPaddings = await this.ctrCipher.decrypt(
			encryptedBuffer.buffer,
			iv,
		);

		const endOfDataOffset = decryptedBufferWithPaddings.byteLength - padding;
		return decryptedBufferWithPaddings.slice(0, endOfDataOffset);
	}
}
