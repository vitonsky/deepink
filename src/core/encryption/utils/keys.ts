import { joinBuffers } from './buffers';

/**
 * Creates and return derived `CryptoKey` from a master password
 */
export async function getMasterKey(
	masterPassword: string | ArrayBuffer,
	salt: ArrayBuffer,
) {
	const codec = new TextEncoder();

	const key =
		typeof masterPassword === 'string'
			? codec.encode(masterPassword)
			: masterPassword;

	return await self.crypto.subtle
		.importKey('raw', key, { name: 'PBKDF2' }, false, ['deriveBits', 'deriveKey'])
		// We don't want to use key from original password, only derived keys,
		// so here we derive a new key based on password
		.then((key) =>
			self.crypto.subtle.deriveKey(
				{
					name: 'PBKDF2',
					salt: new Uint8Array(salt),
					iterations: 100000,
					hash: 'SHA-512',
				},
				key,
				{
					name: 'AES-GCM',
					length: 256,
				},

				// Whether or not the key is extractable
				true,

				// This usage limitations will be ignored, because we import and then export key with new permissions
				['encrypt', 'decrypt'],
			),
		)
		// Export and import of key is necessary, to bypass limitation about key can't be derived of derived key
		// Otherwise generates error `DOMException: key.algorithm does not match that of operation`
		.then((key) => self.crypto.subtle.exportKey('raw', key))
		.then((key) =>
			self.crypto.subtle.importKey('raw', key, { name: 'PBKDF2' }, false, [
				'deriveBits',
				'deriveKey',
			]),
		);
}

type KeyAlgorithm =
	| AlgorithmIdentifier
	| AesDerivedKeyParams
	| HmacImportParams
	| HkdfParams
	| Pbkdf2Params;

/**
 * Creates object to generate derived keys and bytes based on provided master key
 */
export async function getDerivedKeysManager(
	masterKey: CryptoKey,
	salt: Uint8Array<ArrayBuffer>,
) {
	const codec = new TextEncoder();

	const getScopedSalt = (scope: string) =>
		joinBuffers([codec.encode(scope).buffer, salt.buffer]);

	return {
		async getDerivedKey(context: string, algorithm: KeyAlgorithm) {
			return self.crypto.subtle.deriveKey(
				{
					name: 'PBKDF2',
					salt: getScopedSalt(context),
					iterations: 100000,
					hash: 'SHA-512',
				},
				masterKey,
				algorithm,

				// Whether or not the key is extractable (less secure) or not (more secure)
				// when false, the key can only be passed as a web crypto object, not inspected
				true,

				// this web crypto object will only be allowed for these functions
				['encrypt', 'decrypt'],
			);
		},

		async getDerivedBytes(context: string, length: number) {
			if (length % 8 !== 0) throw new TypeError('Length is not multiple to 8');

			return self.crypto.subtle.deriveBits(
				{
					name: 'PBKDF2',
					salt: getScopedSalt(context),
					iterations: 100000,
					hash: 'SHA-256',
				},
				masterKey,
				length,
			);
		},
	};
}
