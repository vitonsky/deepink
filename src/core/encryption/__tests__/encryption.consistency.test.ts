import crc32 from 'crc/calculators/crc32';
import { webcrypto } from 'crypto';

import { AESGCMCipher } from '../ciphers/AES';
import { TwofishCTRCipher } from '../ciphers/Twofish';
import { BufferIntegrityProcessor } from '../processors/BufferIntegrityProcessor';
import { BufferSizeObfuscationProcessor } from '../processors/BufferSizeObfuscationProcessor';
import { PipelineProcessor } from '../processors/PipelineProcessor';
import { getDerivedKeysManager, getMasterKey } from '../utils/keys';

// Returns always the same output, to reproduce encryption output
// Implementation generates a long sequence with no repeats, for transparent fingerprint analyzing
const getRandomBytesMock = (length = 16) =>
	new Uint8Array(length).map(
		(_, idx) => idx + Math.max(0, idx + ((length + idx) % 255)),
	).buffer;

const getKeys = async (parameters: { password: string; salt: string }) => {
	const password = new TextEncoder().encode(parameters.password);
	const salt = new TextEncoder().encode(parameters.salt);

	const derivedKeys = await getMasterKey(password.buffer, salt.buffer).then(
		(masterKey) => getDerivedKeysManager(masterKey, salt),
	);

	const aes = await derivedKeys.getDerivedKey('aes-gcm-cipher', {
		name: 'AES-GCM',
		length: 256,
	});

	const twofish = await derivedKeys
		.getDerivedBytes('twofish-ctr-cipher', 256)
		.then((buffer) => new Uint8Array(buffer));

	return { aes, twofish };
};

// Mock for `crypto` for jest context
beforeEach(() => {
	Object.defineProperties(global, {
		crypto: { value: webcrypto, writable: true },
		self: { value: global, writable: false },
	});
});

test('composed processors returns idempotent result', async () => {
	const keys = await getKeys({
		password: 'SuperSecretPassword',
		salt: 'salt bytes',
	});

	const cipher = new PipelineProcessor([
		new BufferIntegrityProcessor(),
		new BufferSizeObfuscationProcessor(getRandomBytesMock),
		new TwofishCTRCipher(keys.twofish, getRandomBytesMock),
		new AESGCMCipher(keys.aes, getRandomBytesMock),
	]);

	const textSample = 'Hello world! This is encryption example text';
	const encryptedBytes = await cipher.encrypt(
		new TextEncoder().encode(textSample).buffer,
	);
	expect(crc32(new Uint8Array(encryptedBytes))).toMatchSnapshot();

	const decryptedBytes = await cipher.decrypt(encryptedBytes);
	expect(new TextDecoder().decode(decryptedBytes)).toBe(textSample);

	// Encryption for the same text input must be equal, because mocked random generator generates identical sequences
	const encryptedBytes2 = await cipher.encrypt(
		new TextEncoder().encode(textSample).buffer,
	);
	expect(new Uint32Array(encryptedBytes)).toStrictEqual(
		new Uint32Array(encryptedBytes2),
	);

	// Encryption for 2 different texts must not be equal
	const encryptedBytes3 = await cipher.encrypt(
		new TextEncoder().encode('Another text').buffer,
	);
	expect(new Uint32Array(encryptedBytes)).not.toEqual(new Uint32Array(encryptedBytes3));

	// Random bytes decryption must throw exception, because integrity check error
	const randomBytes = getRandomBytesMock(800);
	await expect(cipher.decrypt(randomBytes)).rejects.toThrow();
});
