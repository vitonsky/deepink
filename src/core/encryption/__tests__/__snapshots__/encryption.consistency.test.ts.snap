// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`composed processors returns idempotent result 1`] = `-466581813`;
